File: p733e/flood_fill.js
/**
 * @param {number[][]} image
 * @param {number} sr
 * @param {number} sc
 * @param {number} newColor
 * @return {number[][]}
 */
const DIR = [
  [-1, 0],
  [0, -1],
  [1, 0],
  [0, 1],
];
var floodFill = function (image, sr, sc, newColor) {
  const oldColor = image[sr][sc];
  if (oldColor === newColor) return image;

  let recHelper = (r, c) => {
    image[r][c] = newColor;
    for (let d of DIR) {
      const [nr, nc] = [r + d[0], c + d[1]];
      if (nr < 0 || nr >= image.length) continue;
      if (nc < 0 || nc >= image[r].length) continue;
      if (image[nr][nc] !== oldColor) continue;
      recHelper(nr, nc);
    }
  };
  recHelper(sr, sc);
  return image;
};
// TESTS
[
  {
    image: [
      [1, 1, 1],
      [1, 1, 0],
      [1, 0, 1],
    ],
    sr: 1,
    sc: 1,
    newColor: 2,
  },
  {
    image: [
      [1, 0, 1],
      [0, 1, 0],
      [1, 0, 1],
    ],
    sr: 0,
    sc: 0,
    newColor: 2,
  },
].forEach((t) => {
  console.log(
    'Flood fill',
    t.image,
    'from (' + t.sr + ',' + t.sc + ') by new color',
    t.newColor,
    '->'
  );
  let ans = floodFill(t.image, t.sr, t.sc, t.newColor);
  console.log(ans);
});

File: p733e/flood_fill.py
from typing import List


DIR = [(-1, 0), (0, -1), (1, 0), (0, 1)]


class Solution:
    def floodFill(
        self, image: List[List[int]], sr: int, sc: int, newColor: int
    ) -> List[List[int]]:
        oldColor = image[sr][sc]
        if oldColor == newColor:
            return image
        NR, NC = len(image), len(image[0])

        def recHelper(r: int, c: int) -> None:
            image[r][c] = newColor
            for dr, dc in DIR:
                nr, nc = r + dr, c + dc
                if 0 <= nr < NR and 0 <= nc < NC and image[nr][nc] == oldColor:
                    recHelper(nr, nc)

        recHelper(sr, sc)
        return image


# TESTS
tests = [
    {
        "image": [[1, 1, 1], [1, 1, 0], [1, 0, 1]],
        "sr": 1,
        "sc": 1,
        "newColor": 2,
        "newImage": [[2, 2, 2], [2, 2, 0], [1, 0, 1]],
    },
    {
        "image": [[1, 0, 0], [0, 1, 0], [1, 0, 1]],
        "sr": 0,
        "sc": 0,
        "newColor": 2,
        "newImage": [[2, 0, 0], [0, 1, 0], [1, 0, 1]],
    },
]

for t in tests:
    sol = Solution()
    actual = sol.floodFill(t["image"], t["sr"], t["sc"], t["newColor"])
    print(actual)
